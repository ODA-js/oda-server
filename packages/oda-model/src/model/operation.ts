import clean from '../lib/json/clean';
import { MetaModelType, OperationStorage, OperationInput } from './interfaces';
import { FieldBase } from './fieldbase';

export class Operation extends FieldBase {
  public modelType: MetaModelType = 'operation';
  protected $obj!: OperationStorage;

  public get actionType(): string {
    return this.$obj.actionType;
  }

  public updateWith(obj: OperationInput) {
    if (obj) {
      super.updateWith(obj);

      const result = { ...this.$obj };

      result.actionType = obj.actionType;

      this.$obj = result;
    }
  }

  // it get fixed object
  public toObject() {
    let props = this.$obj;
    let res = super.toObject();
    return clean({
      ...res,
      actionType: props.actionType,
    });
  }

  // it get clean object with no default values
  public toJSON() {
    let props = this.$obj;
    let res = super.toJSON();
    return clean({
      ...res,
      actionType: props.actionType,
    });
  }
}
