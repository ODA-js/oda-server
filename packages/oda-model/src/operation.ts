import {
  ModelBaseInternal,
  IModelBase,
  ModelBaseInput,
  ModelBaseMetaInfo,
  ModelBase,
  ModelBaseOutput,
} from './modelbase';
import {
  IFieldArgs,
  OperationKind,
  AsHash,
  MetaModelType,
  assignValue,
  HashToMap,
  NamedArray,
  ArrayToMap,
  MapToArray,
} from './types';
import { merge } from 'lodash';
import { Internal } from './element';
import decapitalize from './lib/decapitalize';
import { IArgs, Args } from './args';
/**
 * Kind of mutation which is intended to work with single entity
 */
export interface IOperation
  extends IModelBase<OperationMetaInfo, OperationInput, OperationOutput> {
  /**
   * action type
   * CRUD
   */
  readonly actionType: OperationKind;
  /**
   * is it field inherited from other entity and which one
   */
  readonly inheritedFrom?: string;
  /**
   * set of arguments
   */
  readonly args?: Map<string, IArgs>;
  readonly payload: Map<string, IArgs>;
  readonly order: number;
  readonly entity?: string;
}

export interface OperationMetaInfo extends ModelBaseMetaInfo {
  entity: string;
  order: number;
  acl: {
    /** if packages allowed to execute mutation */
    execute: string[];
  };
}

export interface OperationInput extends ModelBaseInput<OperationMetaInfo> {
  args: AsHash<IFieldArgs> | NamedArray<IFieldArgs>;
  payload: AsHash<IFieldArgs> | NamedArray<IFieldArgs>;
  inheritedFrom?: string;
  entity?: string;
  actionType: OperationKind;
  order?: number;
}

export interface OperationOutput extends ModelBaseOutput<OperationMetaInfo> {
  args?: NamedArray<IFieldArgs>;
  inheritedFrom?: string;
  payload: NamedArray<IFieldArgs>;
  entity: string;
  actionType: OperationKind;
  order: number;
}

export interface OperationInternal extends ModelBaseInternal {
  args: Map<string, IArgs>;
  payload: Map<string, IArgs>;
  inheritedFrom?: string;
  entity: string;
  actionType: OperationKind;
}

export const operationDefaultMetaInfo = { acl: { execute: [] } };
export const operationDefaultInput = { metadata: operationDefaultMetaInfo };

export class Operation
  extends ModelBase<
    OperationMetaInfo,
    OperationInput,
    OperationInternal,
    OperationOutput
  >
  implements IOperation {
  public get modelType(): MetaModelType {
    return 'operation';
  }

  public get actionType(): OperationKind {
    return this[Internal].actionType;
  }

  get inheritedFrom(): string | undefined {
    return this[Internal].inheritedFrom;
  }

  get args(): Map<string, IArgs> {
    return this[Internal].args;
  }

  get payload(): Map<string, IArgs> {
    return this[Internal].payload;
  }

  get order(): number {
    return this.metadata.order;
  }

  get entity(): string {
    return this.metadata.entity;
  }

  constructor(init: OperationInput) {
    super(merge({}, operationDefaultInput, init));
  }

  public updateWith(input: OperationInput) {
    super.updateWith(input);

    assignValue<OperationInternal, OperationInput, string>({
      src: this[Internal],
      input,
      field: 'name',
      effect: (src, value) => (src.name = decapitalize(value.trim())),
      required: true,
    });

    assignValue({
      src: this.metadata,
      input,
      field: 'entity',
      effect: (src, value) => (src.entity = value),
    });

    assignValue({
      src: this[Internal],
      input,
      field: 'actionType',
    });

    assignValue({
      src: this.metadata,
      input,
      field: 'order',
      effect: (src, value) => (src.order = value),
    });

    assignValue<
      OperationInternal,
      OperationInput,
      AsHash<IFieldArgs> | NamedArray<IFieldArgs>
    >({
      src: this[Internal],
      input,
      field: 'args',
      effect: (src, value) =>
        (src.args = Array.isArray(value)
          ? ArrayToMap(value, v => new Args(v))
          : HashToMap(value, (name, v) => new Args({ name, ...v }))),
      required: true,
      setDefault: src => (src.args = new Map()),
    });

    assignValue<
      OperationInternal,
      OperationInput,
      AsHash<IFieldArgs> | NamedArray<IFieldArgs>
    >({
      src: this[Internal],
      input,
      field: 'payload',
      effect: (src, value) =>
        (src.payload = Array.isArray(value)
          ? ArrayToMap(value, v => new Args(v))
          : HashToMap(value, (name, v) => new Args({ name, ...v }))),
      required: true,
    });
  }

  public toObject(): OperationOutput {
    return merge({}, super.toObject(), {
      actionType: this.actionType,
      inheritedFrom: this.inheritedFrom,
      args: MapToArray(this[Internal].args, (_name, value) => value.toObject()),
      payload: MapToArray(this[Internal].payload, (_name, value) =>
        value.toObject(),
      ),
    } as Partial<OperationOutput>);
  }
}
