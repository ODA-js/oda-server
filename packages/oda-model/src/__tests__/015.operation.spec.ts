import 'jest';
import { Operation, operationDefaultMetaInfo } from '../operation';

describe('Operation', () => {
  it('default', () => {
    const res = new Operation({
      name: 'updateUser',
      entity: 'User',
      actionType: 'create',
      payload: [{ name: 'result', multiplicity: 'many' }],
      args: [{ name: 'id' }, { name: 'fields', multiplicity: 'many' }],
    });
    expect(res.metadata).toMatchObject(operationDefaultMetaInfo);
    expect(res.toObject()).toMatchSnapshot('toObject');
  });
});
