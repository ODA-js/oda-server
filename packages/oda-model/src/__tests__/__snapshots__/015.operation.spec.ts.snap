// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Operation default actionType: toObject 1`] = `
Object {
  "actionType": "other",
  "args": Array [
    Object {
      "defaultValue": undefined,
      "description": undefined,
      "metadata": Object {},
      "multiplicity": "one",
      "name": "id",
      "required": false,
      "title": undefined,
      "type": "string",
    },
    Object {
      "defaultValue": undefined,
      "description": undefined,
      "metadata": Object {},
      "multiplicity": "many",
      "name": "fields",
      "required": false,
      "title": undefined,
      "type": "string",
    },
  ],
  "description": undefined,
  "inheritedFrom": undefined,
  "metadata": Object {
    "acl": Object {
      "execute": Array [],
    },
    "entity": "User",
  },
  "name": "updateUser",
  "payload": Array [
    Object {
      "defaultValue": undefined,
      "description": undefined,
      "metadata": Object {},
      "multiplicity": "many",
      "name": "result",
      "required": false,
      "title": undefined,
      "type": "string",
    },
  ],
  "title": undefined,
}
`;

exports[`Operation default: toObject 1`] = `
Object {
  "actionType": "create",
  "args": Array [
    Object {
      "defaultValue": undefined,
      "description": undefined,
      "metadata": Object {},
      "multiplicity": "one",
      "name": "id",
      "required": false,
      "title": undefined,
      "type": "string",
    },
    Object {
      "defaultValue": undefined,
      "description": undefined,
      "metadata": Object {},
      "multiplicity": "many",
      "name": "fields",
      "required": false,
      "title": undefined,
      "type": "string",
    },
  ],
  "description": undefined,
  "inheritedFrom": undefined,
  "metadata": Object {
    "acl": Object {
      "execute": Array [],
    },
    "entity": "User",
  },
  "name": "updateUser",
  "payload": Array [
    Object {
      "defaultValue": undefined,
      "description": undefined,
      "metadata": Object {},
      "multiplicity": "many",
      "name": "result",
      "required": false,
      "title": undefined,
      "type": "string",
    },
  ],
  "title": undefined,
}
`;
